File: lintcode/cpp/Search_a_2D_Matrix.cpp
class Solution {
public:
    /**
     * @param matrix, a list of lists of integers
     * @param target, an integer
     * @return a boolean, indicate whether matrix contains target
     */
    bool searchMatrix(vector<vector<int> > &matrix, int target) {
        if(matrix.empty()) return false;
        
        int r = matrix.size();
        int c = matrix[0].size();
        int l = 0;
        int h = r*c-1;
        
        while(l <= h){
            int m = l + (h-l)/2;
            int x = matrix[m/c][m%c];
            
            if(x == target) return true;
            else if(x < target) l = m+1;
            else h = m-1;
        }
        
        return false;
    }
};

File: lintcode/java/Searcha2DMatrix.java
public class Solution {
    /**
     * @param matrix, a list of lists of integers
     * @param target, an integer
     * @return a boolean, indicate whether matrix contains target
     */
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix.length == 0) return false;
        int m = matrix.length;
        int n = matrix[0].length;
        
        int i = 0;
        int j = n-1;
        while(i < m && j >= 0){
            if(target < matrix[i][j]) j--;
            else if(target > matrix[i][j]) i++;
            else return true;
        }
        
        return false;
    }
}

